plutus-core/untyped-plutus-core/Language/UntypedPlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -288,18 +284,23 @@ instance MonadEmitter (CekCarryingM term uni fun s) where
 -- @Ix fun@ which implies @Ord fun@ which implies @Eq fun@.
 instance (Eq fun, Hashable fun, ToExMemory term) =>
             SpendBudget (CekCarryingM term uni fun s) fun (ExBudgetCategory fun) term where
-    spendBudget key budget = do
-        modifying exBudgetStateTally
-                (<> (ExTally (singleton key budget)))
-        newBudget <- exBudgetStateBudget <%= (<> budget)
-        mode <- asks cekEnvBudgetMode
-        case mode of
-            Counting -> pure ()
-            Restricting resb ->
-                when (exceedsBudget resb newBudget) $
-                    throwingWithCause _EvaluationError
-                        (UserEvaluationError $ CekOutOfExError resb newBudget)
-                        Nothing  -- No value available for error
+    spendBudget key budgetSpent = do
+        exSt <- get
+        case exSt of
+            CountingSt budgetTotal -> do
+                let budgetTotal' = budgetTotal <> budgetSpent
+                put $! CountingSt budgetTotal'
+            TallyingSt tallies budgetTotal -> do
+                let budgetTotal' = budgetTotal <> budgetSpent
+                put $! TallyingSt (tallies <> ExTally (singleton key budgetSpent)) budgetTotal'
+            RestrictingSt budgetLeft -> do
+                let budgetLeft' = budgetLeft `minusExBudget` budgetSpent
+                if exceedsBudget budgetLeft budgetSpent
```

Comment:
I got really confused there because I thought that `budgetSpent` was the total amount that we'd spent so far, but in fact it's just the amount we're about to spend right now!  That's just me failing to pay attention though.

Reply:
I'll rename it to `budgetToSpend` (or something else if you have a better suggestion).

> Also, isn't this the same as `not (exceedsBudget budgetLeft' mempty)` (or whatever the zero budget is)

I wanted to avoid negative values in the check and leave them only for the error, but Michael asks the same and I don't really care, so I'll change the check.

Reply:
> I'll rename it to `budgetToSpend`

Done.

> Also, isn't this the same as not (`exceedsBudget budgetLeft' mempty`) (or whatever the zero budget is)

Done (replaced `exceedsBudget` with `isNegativeBudget`).